Add metrics for outer join lowering cases

ggevay opened the pull request:
To aid the planning of work on outer joins, this PR adds a metric that counts each of the cases of outer join lowering.

The first commit just moves `OptimizerMetrics` to `mz_sql`. This is somewhat unfortunate, but I didn't see a better way to make it available in the lowering. I'm open to ideas. (In the long run, I guess it would be great to have an `mz_optimizer` crate, which would include the stuff that is currently in `mz_transform`, and also the lowering, and maybe some other minor things.)

The second commit was a lot of typing, but very straightforward: just wires the `OptimizerMetrics` to every part of the lowering. (I couldn't just wire it to a smaller subset of functions, because all of these many functions are co-recursively calling each other.)

The third commit is the main thing, which actually increments the metric at a couple of places in the lowering.

The fourth commit is a minor tweak to make the code simpler. Before this change, the `case = 1` inside `attempt_left_join_magic` was dead code, because we were avoiding calling `attempt_left_join_magic` altogether when the outer arity is non-0. Removing the check that is before the call to `attempt_left_join_magic` makes the `case = 1` code handle this situation. Note that this very slightly decreases the performance of the lowering, but I think it's worth it to make the code simpler. (Alternatively, we could just have some additional debug logging and metrics incrementing in the else branch of that `if` that was checking the outer arity before calling `attempt_left_join_magic`.)

~I'm thinking about adding a fifth commit, which would further decompose the `voj_4` case. This would be just a local change there, so the rest of the PR is ready for review.~ Edit: There is no time.

### Motivation

  * This PR adds a feature that has not yet been specified. It just adds a metric to help the planning of further work on outer join.

### Tips for reviewer

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

ggevay commented:
I've kicked off a limited Nightly: https://buildkite.com/materialize/nightly/builds/8898
It's SQLSmith, SQLancer, RQG left joins, RQG subqueries.

antiguru commented on src/adapter/src/optimize/materialized_view.rs:
```diff
@@ -168,7 +168,7 @@ impl Optimize<HirRelationExpr> for Optimizer {
         trace_plan!(at: "raw", &expr);
 
         // HIR ⇒ MIR lowering and decorrelation
-        let expr = expr.lower(&self.config)?;
```
The change might be less noisy if the metrics become part of config, or we introduce a (dreadful?) context object that contains both config and metrics?

ggevay replied:
But anyhow it would need to touch every line that the current diff touches.

ggevay replied:
Since we are a bit tight on time for the release cut, I'd prefer keeping it as it is.

teskje replied:
Ah, I commented the same thing below! Fine to do this as a follow-up if we need to rush. But I do think having the meta data encapsulated in a type would also be good for the readability of this code, not just of the diff.

ggevay replied:
Ok, then I'll do this in a follow-up PR right away.

teskje commented on src/sql/src/plan/lowering/variadic_left.rs:
```diff
@@ -364,6 +378,7 @@ pub(crate) fn attempt_left_join_magic(
     Ok(Some(body))
 }
 
+use crate::optimizer_metrics::OptimizerMetrics;
```
The place for this import seems strange, given that the imported type is only used further up in this file.

ggevay replied:
fixed

teskje commented on src/sql/src/plan/lowering.rs:
```diff
@@ -736,11 +736,17 @@ impl HirRelationExpr {
                                     config.enable_new_outer_join_lowering,
                                     config.enable_outer_join_null_filter,
                                 )? {
+                                    metrics.iter().for_each(|metrics| {
```
`metrics` is an `Option`. Is there a reason why you don't use an `if let` here?

(Also applies to the other instances of the same pattern below.)

teskje reviewed (approved):
LGTM, just a couple style comments.

Having to move the optimizer metrics into `mz_sql` is indeed unfortunate but I can't think of a better solution that doesn't involve significant crate refactoring.

ggevay commented on src/adapter/src/optimize/view.rs:
```diff
@@ -17,8 +17,8 @@ use mz_transform::dataflow::DataflowMetainfo;
 
-use crate::optimize::metrics::OptimizerMetrics;
 use crate::optimize::{optimize_mir_local, trace_plan, Optimize, OptimizerConfig, OptimizerError};
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/optimize/subscribe.rs:
```diff
@@ -33,12 +33,12 @@ use crate::optimize::dataflows::{
     OptimizeMode, OptimizerConfig, OptimizerError,
 };
 use crate::CollectionIdBundle;
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/optimize/peek.rs:
```diff
@@ -34,12 +34,12 @@ use crate::optimize::dataflows::{
     OptimizerConfig, OptimizerError,
 };
 use crate::TimestampContext;
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/optimize/materialized_view.rs:
```diff
@@ -45,11 +45,11 @@ use crate::catalog::Catalog;
     optimize_mir_local, trace_plan, LirDataflowDescription, MirDataflowDescription, Optimize,
     OptimizeMode, OptimizerConfig, OptimizerError,
 };
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/optimize/index.rs:
```diff
@@ -43,11 +43,11 @@ use crate::catalog::Catalog;
     trace_plan, LirDataflowDescription, MirDataflowDescription, Optimize, OptimizeMode,
     OptimizerConfig, OptimizerError,
 };
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/optimize/copy_to.rs:
```diff
@@ -38,12 +38,12 @@ use crate::optimize::dataflows::{
     OptimizeMode, OptimizerConfig, OptimizerError,
 };
 use crate::TimestampContext;
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/explain/insights.rs:
```diff
@@ -31,10 +31,10 @@ use serde::Serialize;
 use crate::optimize::{self, Optimize, OptimizerConfig, OptimizerError};
 use crate::session::SessionMeta;
 use crate::TimestampContext;
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented on src/adapter/src/coord.rs:
```diff
@@ -179,13 +179,13 @@ use crate::metrics::Metrics;
 use crate::util::{ClientTransmitter, CompletedClientTransmitter, ResultExt};
 use crate::webhook::{WebhookAppenderInvalidator, WebhookConcurrencyLimiter};
 use crate::{flags, AdapterNotice, ReadHolds, TimestampProvider};
+use mz_sql::optimizer_metrics::OptimizerMetrics;
```
fixed

ggevay commented:
Addressed all comments except for the `Context` encapsulation. I'll do that as a follow-up PR right away, but gonna set auto-merge on this one, because we are tight for the release cut.
